### PR TITLE
Zerryth/qna parity

### LINKED ISSUE #684: QnAMaker trace does not behave the same in JS as it does C#
## Versions
What package version of the SDK are you using. - JS botbuilder-ai 4.1.6
What nodejs version are you using - v10.14.1
What browser version are you using - Chrome

## Describe the bug
PR https://github.com/Microsoft/botbuilder-js/pull/683 implements the use of QnA Maker Trace, as requested by issue #649 by including the `emitTraceInfo()` (already in the JS SDK)

However, 
* JS: the two methods that query QnAMaker service (`generateAnswers()` and `getAnswers()`) only emit trace when question has length greater than 0, 
* C#: whereas in C# the trace is always emitted with every query, including queries with null or empty questions

Additionally `QnAMakerTrace` should also be changed to include new parameter values for `strictFilters` and `metadataBoost` -- currently both values are only set to empty `[{}]`

## To Reproduce
query QnA service (bug reproduces in both new `getAnswers()` and deprecated `generateAnswer()` with empty question & note no qna maker trace

## Expected behavior
For trace to be emitted with all queries as in C#

### PR DESCRIPTION
**Pull request by Zerryth**
Fixes #684, #676.
Also addresses concerns outlined in #554 (cleaner code factoring, unit tests, configuration, tracing, options)

## Description
- Now throws error outlining that a `QnAMakerEndpoint` is needed to instantiate `QnAMaker`
- Trace now emits on all calls to qna in `getAnswers()`
- `emitTraceInfo()` now correctly includes both options that are passed in during `QnAMaker` instantiation _as well as_ any `QnAMakerOptions` passed into call to qna service in `getAnswers()` 
  - Note to being along the lines of parity: the above is a bug that still needs to be fixed on the C# side 
- bunch of unit tests

## Specific Changes
  - constructor of `QnAMaker` throws error if no `QnAMakerEndpoint`
  - As in C# side, added `validateOptions()` method
    - Throws helpful errors needed to send valid request to qna such as for `scoreThreshold` (`must have value between 0 and 1`) and `top` (`must be an integer greater than 0`)
  - Updated documentation on `strictFilters` and `metadataBoost`
  - refactored `.getAnswers()` and its helper methods to not break on unit test where `undefined` is passed in
  - exercised clean coding practice of abstracting lower-level processes in `.getAnswers()` and `.queryQnaService()` into helper functions to make code more expressive/readable and better organized
    - **`getAnswers()`**: `getTrimmedMessageText()`, `validateOptions()`, `sortAnswersWithinThreshold()`
    - **`queryQnaService()`**: `getHeaders()`, `formatQnaResult()`
 - extracting user's `question` in `getAnswers()` and deprecated `generateAnswers()` now has check to ensure that `context` has `message` type before trying to access the `text` using `getTrimmedMessageText()`
- `emitTraceInfo()` signature changed to accept `queryOptions` to be able to properly include options passed into call and not just options set in the `QnAMaker` constructor 
- moved deprecated methods towards the bottom of `QnAMaker` class file

## Testing
- beefed up code coverage for `qnaMaker.ts`

### CODE REVIEW

**Comment by Zerryth** [comment]
Steven sent me the Details dump that I currently don't have permission to see myself.

And hmm...it looks like several of the QnAMaker tests are breaking with error "TypeError: expected.test is not a function" and "assert.rejects is not a function"

The latter is confusing, because `.rejects()` is definitely a part of Node's assert library. Digging...

**Comment by Zerryth** [comment]
Alright, working with build engineer Jose on the failing issue.

Tests are passing in Travis, but not in BotBuilder-JS master build, because:

"Azure Devops, due to the locked-down nature of our repositories, is preventing the test build from accessing outside resources - such as QnAMaker."

Investigating, you can see in "Details" that it breaks on unit tests that specifically make a call to QnAMaker service. 

Pending Jose's help to see if we can get the connecting-to-the-service issue resolved

**Comment by cleemullins** [comment]
@Zerryth Hey Ashley, is this ready to merge?

**Comment by Zerryth** [comment]
@cleemullins let me talk with Jose regarding the errors he's seeing on why build is failing

I think @stevengum / @johnataylor still haven't reviewed it yet either, because "some checks were not successful" status

**Comment by godsanchez** [comment]
Re-opening to re-trigger build tests

**Comment by Zerryth** [comment]
YAAAAS!!

Ok. We figured out the build issue!
* Basically AzureDevOps build was using older version of node, when the some of the assertion library methods in node that are used in tests to test async code were actually added in `v10.0.0`
* And most likely travis was using a more updated version, which was why tests were passing there, even though they failed in Azure
___
#### @stevengum feel free to have a gander now to review

**Comment by stevengum** [comment]
@Zerryth I still need to review this, but there is an errant qnamaker.js file in the src folder. Please remove [this](https://github.com/Microsoft/botbuilder-js/pull/704/files#diff-64a1c5a084b71e4186c7e7bbfd359526) 😄

**Comment by Zerryth** [comment]
Removed the sneaky .js file!

**Comment by cleemullins** [comment]
@stevengum, @johnataylor Can you guys look at this? Let's get this in. Too much good stuff here to leave this PR hanging.

File: libraries/botbuilder-ai/tests/qnaMaker.test.js [hunk 1]
```diff
@@ -70,111 +71,298 @@ describe('QnAMaker', function () {
...
-    });
+        it('should instantiate a QnAMaker class successfully with both QnAMakerEndpoint and QnAMakerOptions args', function() {
+            const options = { top: 7 };
+            const qnaWithOptions = new QnAMaker(endpoint, options);
+            const isQnaMaker = qnaWithOptions instanceof QnAMaker;
 
-    it('should return 0 answers for a question with no answer after a succesful call', async function () {
-        const qna = new ai.QnAMaker(endpoint, { top: 1 });
-        let answer = 'BaseCamp: You can use a damp rag to clean around the Power Pack';
+            assert.strictEqual(isQnaMaker, true, 'Should create QnAMake instance successfully with QnAMakerOptions.');
```
Comment (stevengum) [inline libraries/botbuilder-ai/tests/qnaMaker.test.js]: While this wouldn't be available in TypeScript, in JavaScript you can evaluate the value of `qnaWithOptions._options` to verify that the options were successfully passed in. The feedback on the previous test is applicable to here as well.

File: libraries/botbuilder-ai/src/qnaMaker.ts [hunk 2]
```diff
@@ -365,4 +313,159 @@ export class QnAMaker {
...
+
+    /**
+     * Calls [generateAnswer()](#generateanswer) and sends the resulting answer as a reply to the user.
+     *
+     * @remarks
+     * Returns a value of `true` if an answer was found and sent. If multiple answers are
+     * returned the first one will be delivered.
+     * @param context Context for the current turn of conversation with the user.
+     */
+    public async answer(context: TurnContext): Promise<boolean> {
```
[the remark this answers is not in the document]
Comment (Zerryth) [reply libraries/botbuilder-ai/src/qnaMaker.ts]: Well. I sort of "half-wayed" this here. 
Half-way'd as in:
* I moved the method to the bottom of the class with the deprecated method
* but I didn't actually go through with decorating it `@deprecated`

You and I verbally agreed that yeah, `.answers()` should be deprecated, however your [comment from different qna PR](https://github.com/Microsoft/botbuilder-js/pull/649#discussion_r236531377) went unanswered.

You asked:

> I think we should deprecate this method as well. There is no corresponding AnswerAsync() in dotnet, and a bot should always examine/store the answer before sending it to the user. I have some other misgivings around the method as well...
> 
> @vishwacsena, @cleemullins, what do you think in regards to marking this as deprecated?

[the remark this answers is not in the document]
Comment (Zerryth) [reply libraries/botbuilder-ai/src/qnaMaker.ts]: I guess, should we follow through with actually deprecating `.answers()`? @cleemullins @vishwacsena @stevengum

[the remark this answers is not in the document]
Comment (stevengum) [reply libraries/botbuilder-ai/src/qnaMaker.ts]: I'm for deprecating `answers()` mainly for the reason that a bot should not get answers and then send it to the user without examining it (and the current state) and determining if sending the answers is the best course of action

**Comment by Zerryth** [comment]
Ok, applied changes Steven requested in review.

Now blocked by #727 issue. Mr. Gum requests that we fix the linting before we merge the PR to ensure the order of methods in the `QnAMaker` class is clean

**Pull request merged**